adapter/service/mediaplayer/service.go

```diff
@@ -0,0 +1,320 @@
+	SendPlaybackModeReport(force bool) (bool, error)
+	SetVolume(level int64) error
+	SendVolumeReport(force bool) (bool, error)
+	SetMute(mute bool) error
+	SendMuteReport(force bool) (bool, error)
+	SendMetadataReport(force bool) (bool, error)
+}
+
+type (
+	// PlaybackAction represents a playback action.
+	PlaybackAction string
+	// PlaybackMode represents a playback mode.
+	PlaybackMode struct {
+		Repeat    bool `mapstructure:"repeat"`
+		RepeatOne bool `mapstructure:"repeat_one"`
+		Shuffle   bool `mapstructure:"shuffle"`
+		CrossFade bool `mapstructure:"crossfade"`
+	}
+
+	// Metadata represents the metadata of the media.
+	Metadata struct {
+		Album    string `json:"album"`
+		Track    string `json:"track"`
+		Artist   string `json:"artist"`
+		ImageURL string `json:"image_url"`
+	}
+
+	// Config represents a service configuration.
+	Config struct {
+		Specification     *fimptype.Service
+		Controller        Controller
+		ReportingStrategy cache.ReportingStrategy
+	}
+)
+
+// NewService creates a new instance of a mediaplayer FIMP service.
+func NewService(
+	publisher adapter.ServicePublisher,
+	cfg *Config,
+) Service {
+	cfg.Specification.EnsureInterfaces(requiredInterfaces()...)
+
+	if cfg.ReportingStrategy == nil {
+		cfg.ReportingStrategy = DefaultReportingStrategy
+	}
+
+	return &service{
+		Service:           adapter.NewService(publisher, cfg.Specification),
+		controller:        cfg.Controller,
+		reportingStrategy: cfg.ReportingStrategy,
+
+		reportingCache: cache.NewReportingCache(),
+		lock:           &sync.Mutex{},
+	}
+}
+
+// service is a private implementation of a mediaplayer FIMP service.
+type service struct {
+	adapter.Service
+
+	controller        Controller
+	lock              *sync.Mutex
+	reportingCache    cache.ReportingCache
+	reportingStrategy cache.ReportingStrategy
+}
+
+// SetPlayback sets the playback state.
+func (s service) SetPlayback(action PlaybackAction) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if err := action.Validate(); err != nil {
+		return fmt.Errorf("%s: invalid playback action: %w", s.Name(), err)
+	}
+
+	err := s.controller.SetPlayback(action)
+	if err != nil {
+		return fmt.Errorf("%s: failed to set playback: %w", s.Name(), err)
+	}
+
+	return nil
+}
+
+// SendPlaybackReport sends a playback report. Returns true if a report was sent.
+func (s service) SendPlaybackReport(force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	action, err := s.controller.Playback()
+	if err != nil {
+		return false, fmt.Errorf("%s: failed to get playback report: %w", s.Name(), err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtPlaybackReport, "", action) {
+		return false, nil
+	}
+
+	message := fimpgo.NewStringMessage(EvtPlaybackReport, s.Name(), string(action), nil, nil, nil)
+
+	err = s.SendMessage(message)
+	if err != nil {
+		return false, fmt.Errorf("%s: failed to send playback report: %w", s.Name(), err)
+	}
+
```

Comment:
Please call `reportingCache.Reported` here and in all the methods where you call `ReportRequired` method.